File: sources/java/_05_alloc_churn.java
// task 05 alloc_churn — expected output: 1274991808
// build: javac _05_alloc_churn.java    run: java _05_alloc_churn
// build (graalvm native-image): native-image -O2 _05_alloc_churn    run: ./_05_alloc_churn
// The 256 slots keep each buffer reachable; the array it replaces becomes garbage for the GC.

public class _05_alloc_churn {
    public static void main(String[] args) {
        long total = 0;
        byte[][] slots = new byte[256][];
        for (long i = 0; i < 10000000L; i++) {
            int v = (int) (i % 256);
            byte[] buf = new byte[64];
            buf[0] = (byte) v;
            total += v;
            slots[v] = buf;
        }
        System.out.println(total);
    }
}

File: sources/php/05_alloc_churn.php
<?php
// task 05 alloc_churn — expected output: 1274991808
// build: none (interpreted)    run: php 05_alloc_churn.php (zend) | php -d opcache.enable_cli=1 -d opcache.jit_buffer_size=64M 05_alloc_churn.php (zend + jit)
// A PHP string is a mutable byte array, so each iteration allocates one 64-byte buffer;
// the slot store keeps it reachable and drops the buffer it replaces.

$total = 0;
$slots = array_fill(0, 256, '');

for ($i = 0; $i < 10000000; $i++) {
    $buf = str_repeat("\0", 64);
    $buf[0] = chr($i & 255);
    $total += ord($buf[0]);
    $slots[$i & 255] = $buf;
}

echo $total, "\n";
